src/cudamatrix/cu-kernels.cu

```diff
@@ -1641,6 +1803,172 @@ static void _vec_transform_reduce(
     result[blockIdx.x] = op.PostReduce(sdata[0], result[blockIdx.x]);
 }
 
+
+template <typename T, typename ReduceLambda>
+inline __device__ void myAtomicReduce(T *address, T val, ReduceLambda op);
+
+template <typename ReduceLambda>
+inline __device__ void myAtomicReduce(double *address, double val, ReduceLambda op) {
+  unsigned long long int *address_as_ull = (unsigned long long int *)address;
+  unsigned long long int old = *address_as_ull, assumed;
+  do {
+    assumed = old;
+    old =
+      atomicCAS(address_as_ull, assumed,
+                __double_as_longlong(op.Reduce(val, __longlong_as_double(assumed))));
+  } while (assumed != old);
+}
+
+template <typename ReduceLambda>
+inline __device__ void myAtomicReduce(float *address, float val, ReduceLambda op) {
+  unsigned int *address_as_uint = (unsigned int *)address;
+  unsigned int old = *address_as_uint, assumed;
+  do {
+    assumed = old;
+    old = atomicCAS(address_as_uint, assumed,
+                    __float_as_uint(op.Reduce(val, __uint_as_float(assumed))));
+  } while (assumed != old);
+}
+
+template <>
+inline __device__ void myAtomicReduce(double *address, double val, TransReduceOp<SUMAB, double> op) {
+  myAtomicAdd(address, val);
+}
+
+template <>
+inline __device__ void myAtomicReduce(float *address, float val, TransReduceOp<SUMAB, float> op) {
+  myAtomicAdd(address, val);
+}
+
+// Reduce a matrix 'data' to a row vector 'dots'
+template <EnumTransformReduce TransReduceType, typename Real, int unroll_count>
+__global__ void _strided_reduction_fused_kernel(Real * __restrict__ dots, const Real * __restrict__ data,
+                                                void * __restrict__ scratch, const MatrixDim d,
```

Comment:
Why is scratch `void*` and not `Real`?

Reply:
`scratch` is a workspace that is split into two portions. One to store the intermediate results of type `Real` and the other to store `unsigned int` flags to guarantee ordering of operations across cuda blocks. Since all of this is internal to the kernel, I thought it made more sense to expose it as a single `void *` argument rather than have two separate `Real *` and `unsigned int *` arguments.